### PR TITLE
Save assets in a state file on disk

### PR DESCRIPTION
**Pull request by rajatchopra**
pkg/asset:

Implement new funtions Save/Load for saving and loading the asset states

cmd/openshift-install:

Call out on the Save function to persist the state file of all assets in a constant state file

### CODE REVIEW

File: pkg/asset/store.go [hunk 1]
```diff
@@ -24,6 +32,36 @@ func (s *StoreImpl) Fetch(asset Asset) (*State, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(data, &assets)
+	return assets, err
+}
+
+// Save dumps the entire state map into a file
+func (s *StoreImpl) Save(dir string) error {
+	data, err := yaml.Marshal(&s.assets)
```
Comment (crawford) [inline pkg/asset/store.go]: Let's use JSON instead. Go's YAML implementation has a lot of trouble serializing and then deserializing accurately and YAML is a little too inviting for end users.

**Comment by rajatchopra** [comment]
Have to redo the PR. Closing this one out.

**Comment by crawford** [comment]
Please don't close out these pull requests. There is valuable discussion (maybe not so much in this case) that gets lost. Unless you are fundamentally changing the approach, I would reuse existing pull requests.

**Comment by yifan-gu** [comment]
@rajatchopra Any updates on this one?

**Comment by rajatchopra** [comment]
@yifan-gu @staebler PTAL
This PR uses Name() for the keys, and I can successfully marshal/unmarshal the asset map. Do we actually ever need to restore the asset objects? I am guessing no. Good riddance.

Try this PR with a fresh install. Then repeat without deleting the target directory. No more asset re-generation.

File: cmd/openshift-install/main.go [hunk 2]
```diff
@@ -100,6 +100,11 @@ func main() {
 				logrus.Fatalf("Failed to write asset (%s) to disk: %v", asset.Name(), err)
 			}
 		}
+		err := assetStore.Save(*dirFlag)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/main.go]: This leaves the state file save if we error out in generating any assets above.

File: cmd/openshift-install/main.go [hunk 3]
```diff
@@ -100,6 +100,11 @@ func main() {
...
+		if err != nil {
+			logrus.Fatalf("Failed to write to state file: %v", err)
+			os.Exit(1)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/main.go]: Fatal already os.Exit(1)

File: pkg/asset/stock/stock.go [hunk 4]
```diff
@@ -70,3 +70,15 @@ func EstablishStock() *Stock {
 	s.metadataStock.EstablishStock(s, s)
 	return s
 }
+
+//
+// func AssetConstructor(typeof string) (Asset, err) {
```
Comment (abhinavdahiya) [inline pkg/asset/stock/stock.go]: ??

File: pkg/asset/store.go [hunk 5]
```diff
@@ -1,10 +1,19 @@
...
 	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
 	"github.com/sirupsen/logrus"
 )
 
+const (
+	stateFileName = ".assetState.json"
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: can we not camel case. maybe this: `.openshift_install_state.json`

File: pkg/asset/store.go [hunk 6]
```diff
@@ -23,6 +32,37 @@ func (s *StoreImpl) Fetch(asset Asset) (*State, error) {
...
+		return nil, err
+	}
+	err = json.Unmarshal(data, assets)
+	return assets, err
+}
+
+// Save dumps the entire state map into a file
+func (s *StoreImpl) Save(dir string) error {
+	data, err := json.MarshalIndent(&s.assets, "", "    ")
+
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: drop this extra line.

File: pkg/asset/userprovided.go [hunk 7]
```diff
@@ -24,7 +24,8 @@ func (a *UserProvided) Dependencies() []Asset {
 }
 
 // Generate queries for input from the user.
-func (a *UserProvided) Generate(map[Asset]*State) (state *State, err error) {
+func (a *UserProvided) Generate(map[string]*State) (*State, error) {
```
Comment (abhinavdahiya) [inline pkg/asset/userprovided.go]: this change seems orthogonal.

File: pkg/asset/store.go [hunk 8]
```diff
@@ -34,15 +34,21 @@ func (s *StoreImpl) Fetch(asset Asset) (*State, error) {
...
 // Load retrieves the state from the state file present in the given directory
 // and returns the assets map
-func (s *StoreImpl) Load(dir string) (map[string]*State, error) {
+func (s *StoreImpl) Load(dir string) error {
 	path := filepath.Join(dir, stateFileName)
 	assets := make(map[string]*State)
 	data, err := ioutil.ReadFile(path)
 	if err != nil {
-		return nil, err
+		if _, ok := err.(*os.PathError); ok {
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: https://golang.org/pkg/os/#IsNotExist

File: pkg/asset/userprovided.go (hunk 7 above)
Comment (rajatchopra) [reply pkg/asset/userprovided.go]: Only part of it. The map in the argument does need to change. Fixed.

File: pkg/asset/store.go (hunk 8 above)
Comment (rajatchopra) [reply pkg/asset/store.go]: Fixed.

File: pkg/asset/store.go (hunk 6 above)
Comment (rajatchopra) [reply pkg/asset/store.go]: Ack.

File: pkg/asset/store.go (hunk 5 above)
Comment (rajatchopra) [reply pkg/asset/store.go]: .openshift_install_state.json now.

File: pkg/asset/stock/stock.go (hunk 4 above)
Comment (rajatchopra) [reply pkg/asset/stock/stock.go]: rebase relic :). Some similar code might be if we ever need to be able to construct the asset object from its name. Removed.

File: pkg/asset/store.go [hunk 9]
```diff
@@ -23,16 +32,52 @@ func (s *StoreImpl) Fetch(asset Asset) (*State, error) {
...
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	err = json.Unmarshal(data, &assets)
+	if err == nil {
+		s.assets = assets
```
Comment (yifan-gu) [inline pkg/asset/store.go]: Hmm, not sure we want to overwrite the in-memory store with the on-disk state file, because we discussed that an asset should always fetch its dependencies to determine whether its dependencies are on disk, if so the asset needs to be regenerated.
https://coreos.slack.com/archives/C68TNFWA2/p1538618620000100

Comment (yifan-gu) [reply pkg/asset/store.go]: cc @staebler ^

File: pkg/asset/tls/certkey.go [hunk 10]
```diff
@@ -75,6 +75,8 @@ func (c *CertKey) Generate(parents map[asset.Asset]*asset.State) (*asset.State,
 			if err != nil {
 				return nil, errors.Wrap(err, "failed to generate Subject")
 			}
+			// assign it to the asset object also so that Name() can be generated correctly
+			c.Subject = cfg.Subject
```
[the remark this answers is not in the document]
Comment (staebler) [reply pkg/asset/tls/certkey.go]: >  we must be aware that the type approach forces that we have to have singleton objects of each asset type

This is already required by #344. The key of the asset map in the store is the type.

File: pkg/asset/store.go (hunk 9 above)
Comment (staebler) [reply pkg/asset/store.go]: Agreed. I made a similar argument in my review.

File: cmd/openshift-install/targets.go [hunk 11]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/openshift/installer/pkg/asset/kubeconfig"
 	"github.com/openshift/installer/pkg/asset/manifests"
 	"github.com/openshift/installer/pkg/asset/metadata"
+	"github.com/openshift/installer/pkg/asset/util"
```
Comment (abhinavdahiya) [inline cmd/openshift-install/targets.go]: i don't see this package addition in this commit.
Also please don't add util packages.

File: cmd/openshift-install/targets.go [hunk 12]
```diff
@@ -70,6 +75,10 @@ func runTargetCmd(targets ...asset.WritableAsset) func(cmd *cobra.Command, args
 				return errors.Wrapf(err, "failed to write asset (%s) to disk", a.Name())
 			}
 		}
+		err = assetStore.Save(*dirFlag)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/targets.go]: No state is saved when we fail to generate am asset... ?

File: pkg/asset/store.go [hunk 13]
```diff
@@ -1,12 +1,20 @@
...
+	"os"
+	"path/filepath"
 	"reflect"
 
-	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+const (
+	stateFileName = ".asset_state.json"
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: if somebody looks at this file, they wouln't know it belongs to openshift install.

File: pkg/asset/store.go [hunk 14]
```diff
@@ -25,6 +33,41 @@ func (s *StoreImpl) Fetch(asset Asset) error {
 	return s.fetch(asset, "")
 }
 
+// Load retrieves the state from the state file present in the given directory
+// and returns the assets map
+func (s *StoreImpl) Load(dir string) error {
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: this function doesn't seem to be doing anything ? it loads the state where ?

File: pkg/asset/store.go [hunk 15]
```diff
@@ -25,6 +33,41 @@ func (s *StoreImpl) Fetch(asset Asset) error {
...
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(data, &assets)
+}
+
+// Save dumps the entire state map into a file
+func (s *StoreImpl) Save(dir string) error {
+	assetMap := make(map[string][]*File)
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: This seems contradictory to what @staebler recommended. the state file stores the actual type not just files.

File: pkg/asset/asset.go [hunk 16]
```diff
@@ -26,6 +26,9 @@ type WritableAsset interface {
 
 	// Files returns the files to write.
 	Files() []*File
+
+	// SetState sets the asset data from an unmarshalled File slice
+	SetState([]*File) error
```
Comment (abhinavdahiya) [inline pkg/asset/asset.go]: What is the reason to add another function to the interface. Can you document it.

File: pkg/asset/cluster/cluster.go [hunk 17]
```diff
@@ -102,3 +102,9 @@ func (c *Cluster) Generate(parents asset.Parents) error {
 func (c *Cluster) Files() []*asset.File {
 	return []*asset.File{c.file}
 }
+
+// SetState sets the appropriate fields from given unmarshalled files
+func (c *Cluster) SetState(files []*asset.File) error {
```
Comment (abhinavdahiya) [inline pkg/asset/cluster/cluster.go]: This approach breaks all the code where we were using the asset to get information not reading the file like this https://github.com/openshift/installer/blob/master/pkg/asset/manifests/operators.go#L175

File: pkg/asset/util/util.go [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests"
+	"github.com/openshift/installer/pkg/asset/tls"
+)
+
+// GetTypeFromString takes the asset type as string and returns an actual instance of that type
+func GetTypeFromString(assetType string) (asset.WritableAsset, error) {
```
Comment (abhinavdahiya) [inline pkg/asset/util/util.go]: Why is this required?

File: pkg/asset/manifests/network-operator.go [hunk 19]
```diff
@@ -15,29 +15,29 @@ const (
 	defaultMTU = "1450"
 )
 
-// networkOperator generates the network-operator-*.yml files
-type networkOperator struct {
+// NetworkOperator generates the network-operator-*.yml files
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/network-operator.go]: Why make it public?

File: cmd/openshift-install/targets.go (hunk 11 above)
Comment (rajatchopra) [reply cmd/openshift-install/targets.go]: It can be called something else, but we need a helper/util function outside of asset package to be able to unmarshal strings to actual type.

File: pkg/asset/store.go (hunk 15 above)
Comment (rajatchopra) [reply pkg/asset/store.go]: It does store the actual type only. In the form of a string.

File: pkg/asset/asset.go (hunk 16 above)
Comment (rajatchopra) [reply pkg/asset/asset.go]: Ack. Need to document the reason properly. Just to jot down here though:
We need a function that can restore the content of the WritableAsset when it has been unmarshalled from the state file. Since we unmarshal using a dynamic interface, this is a form of custom unmarshal based on object type.

File: pkg/asset/cluster/cluster.go (hunk 17 above)
Comment (rajatchopra) [reply pkg/asset/cluster/cluster.go]: We can only marshal/unmarshal WritableAssets. If there is a non-writable-asset dependency, then 'generate' will be called for the 'dirty' tree branch anyway.

File: pkg/asset/util/util.go (hunk 18 above)
Comment (rajatchopra) [reply pkg/asset/util/util.go]: Because we dump a map where the type is the key. And when we unmarshal we cannot dynamically construct the object in golang. The primary type of Asset is interface{}, if it were a struct we could have made a custom json unmarshal function. This is the most compact way I can think of in the current scenario.

File: pkg/asset/manifests/network-operator.go (hunk 19 above)
Comment (rajatchopra) [reply pkg/asset/manifests/network-operator.go]: Just so that we can unmarshal it from outside the package. Its type is 'manifests.networkOperator' and to be able to call its constructor for unmarshal we need it to be exported.

File: pkg/asset/util/util.go (hunk 18 above)
Comment (abhinavdahiya) [reply pkg/asset/util/util.go]: https://golang.org/pkg/reflect/#TypeOf ? why can't we use this.

File: pkg/asset/cluster/cluster.go (hunk 17 above)
Comment (abhinavdahiya) [reply pkg/asset/cluster/cluster.go]: When loading installconfig asset you only load the files. but the code https://github.com/openshift/installer/blob/master/pkg/asset/manifests/operators.go#L175 uses the embedded `type.installconfig` directly not the file (which we want to continue to support).

This new SetState doesn't allow that.

File: pkg/asset/store.go (hunk 15 above)
Comment (abhinavdahiya) [reply pkg/asset/store.go]: Can you start documentation why we chose `The non-writable assets do not go into the state file`.

File: pkg/asset/util/util.go (hunk 18 above)
Comment (rajatchopra) [reply pkg/asset/util/util.go]: Can't use TypeOf, because all we have is the string of the Type. reflect.Type cannot be put into json unless we transform it to a string. How do we get the actual type from the 'string'? We have to hardcode a typemap.
For this exact reason, this PR last week had the asset.Name() as the key so that we do not need to implement this hardcoded mapper. It was brought down because Name() will be forced to be unique && something that cannot be generated at runtime. (e.g. certkey).

File: pkg/asset/asset.go (hunk 16 above)
Comment (rajatchopra) [reply pkg/asset/asset.go]: We can do the delayed unmarshal. That was one of the ideas that crossed my mind.
Another one was not to unmarshal the asset object itself. Just store the map[type.String][]*File just like that. And any asset can look up for its 'files' by indexing on its own type.String().

However, we would still need the constructor to be public and some method to load up the object's contents from outside (as I expect this would typically be used in the storeImpl.fetch() method).

File: pkg/asset/util/util.go (hunk 18 above)
Comment (abhinavdahiya) [reply pkg/asset/util/util.go]: https://golang.org/pkg/reflect/#Type
```go
        // Name returns the type's name within its package for a defined type.
        // For other (non-defined) types it returns the empty string.
        Name() string
```

File: pkg/asset/cluster/cluster.go (hunk 17 above)
Comment (abhinavdahiya) [reply pkg/asset/cluster/cluster.go]: https://github.com/openshift/installer/pull/388/files#diff-74e3d89f2a13e465091f905f65cfc646R144 doesn't seem to be doing what you are saying...

Comment (rajatchopra) [reply pkg/asset/cluster/cluster.go]: Yes, you are right. Its still a TODO :). Sorry if that confused you.. I should have put in a comment there.

File: pkg/asset/asset.go (hunk 16 above)
Comment (abhinavdahiya) [reply pkg/asset/asset.go]: > We can do the delayed unmarshal.

It will allow us to no do this SetState extra functions. I would rather do this than add SetState on each asset.

File: pkg/asset/util/util.go (hunk 18 above)
Comment (rajatchopra) [reply pkg/asset/util/util.go]: When I tested this, all our object types return the empty string. Some valid, but undiscovered theory there.

Comment (abhinavdahiya) [reply pkg/asset/util/util.go]: All the assets are pointers. That's why they have empty return from Name....

Check doc for Elem to get underlying type for pointer.

File: pkg/asset/asset_test.go [hunk 20]
```diff
@@ -26,11 +26,11 @@ func (a *persistAsset) Generate(Parents) error {
 
 type writablePersistAsset struct {
 	persistAsset
-	files []*File
+	FileList []*File
```
Comment (abhinavdahiya) [inline pkg/asset/asset_test.go]: `files` seems like the correct naming.

File: pkg/asset/ignition/machine/master.go [hunk 21]
```diff
@@ -14,8 +14,8 @@ import (
 
 // Master is an asset that generates the ignition config for master nodes.
 type Master struct {
-	configs []*igntypes.Config
-	files   []*asset.File
+	Configs  []*igntypes.Config
+	FileList []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/machine/master.go]: again `Files` is more appropriate.

File: pkg/asset/installconfig/installconfig.go [hunk 22]
```diff
@@ -89,19 +89,19 @@ func (a *InstallConfig) Generate(parents asset.Parents) error {
...
 	numberOfWorkers := int64(3)
 	switch {
-	case platform.aws != nil:
-		a.Config.AWS = platform.aws
-	case platform.openstack != nil:
-		a.Config.OpenStack = platform.openstack
-	case platform.libvirt != nil:
-		a.Config.Libvirt = platform.libvirt
-		a.Config.Libvirt.Network.Name = clusterName.clusterName
+	case platform.Aws != nil:
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/installconfig.go]: `AWS`

File: pkg/asset/installconfig/platform.go [hunk 23]
```diff
@@ -58,9 +58,9 @@ var (
 // Platform is an asset that queries the user for the platform on which to install
 // the cluster.
 type platform struct {
-	aws       *types.AWSPlatform
-	openstack *types.OpenStackPlatform
-	libvirt   *types.LibvirtPlatform
+	Aws       *types.AWSPlatform
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/platform.go]: `AWS`

File: pkg/asset/manifests/network-operator.go [hunk 24]
```diff
@@ -15,29 +15,29 @@ const (
...
 )
 
-// networkOperator generates the network-operator-*.yml files
-type networkOperator struct {
-	config *tectonicnetwork.OperatorConfig
-	files  []*asset.File
+// NetworkOperator generates the network-operator-*.yml files
+type NetworkOperator struct {
+	config   *tectonicnetwork.OperatorConfig
+	FileList []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/network-operator.go]: `Files`

File: pkg/asset/manifests/network-operator.go (hunk 19 above)
Comment (abhinavdahiya) [reply pkg/asset/manifests/network-operator.go]: @rajatchopra any thoughts on keeping the struct private?

File: pkg/asset/manifests/operators.go [hunk 25]
```diff
@@ -25,9 +25,9 @@ const (
 
 // Manifests generates the dependent operator config.yaml files
 type Manifests struct {
-	kubeSysConfig  *configurationObject
-	tectonicConfig *configurationObject
-	files          []*asset.File
+	KubeSysConfig  *configurationObject
+	TectonicConfig *configurationObject
+	FileList       []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/operators.go]: `Files`

File: pkg/asset/metadata/metadata.go [hunk 26]
```diff
@@ -87,3 +87,9 @@ func (m *Metadata) Generate(parents asset.Parents) error {
 func (m *Metadata) Files() []*asset.File {
 	return []*asset.File{m.file}
 }
+
+// SetState sets the appropriate fields from given unmarshalled files
```
Comment (abhinavdahiya) [inline pkg/asset/metadata/metadata.go]: This seems to be not required anymore.

File: pkg/asset/store.go [hunk 27]
```diff
@@ -1,12 +1,20 @@
...
+	"os"
+	"path/filepath"
 	"reflect"
 
-	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+const (
+	stateFileName = ".openshift_asset_state.json"
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: `.openshift_install_state.json` / `openshift_install.state` seems more appropriate.

File: pkg/asset/store.go [hunk 28]
```diff
@@ -1,12 +1,20 @@
 package asset
 
 import (
+	"encoding/json"
+	"github.com/pkg/errors"
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: It seems like most of the places we use this grouping
```
<std lib imports>

<external imports>

<imports from this repo>
```

File: pkg/asset/tls/certkey.go [hunk 29]
```diff
@@ -32,19 +32,19 @@ type CertKeyInterface interface {
 // CertKey contains the private key and the cert that's
 // signed by the parent CA.
 type CertKey struct {
-	cert  []byte
-	key   []byte
-	files []*asset.File
+	certBytes []byte
+	keyBytes  []byte
+	FileList  []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/tls/certkey.go]: `Files`

File: pkg/asset/tls/certkey.go [hunk 30]
```diff
@@ -32,19 +32,19 @@ type CertKeyInterface interface {
 // CertKey contains the private key and the cert that's
 // signed by the parent CA.
 type CertKey struct {
-	cert  []byte
-	key   []byte
-	files []*asset.File
+	certBytes []byte
```
Comment (abhinavdahiya) [inline pkg/asset/tls/certkey.go]: `certRaw` or just `cert`

File: pkg/asset/tls/keypair.go [hunk 31]
```diff
@@ -15,9 +15,9 @@ type KeyPairInterface interface {
 
 // KeyPair contains a private key and a public key.
 type KeyPair struct {
-	private []byte
-	public  []byte
-	files   []*asset.File
+	pvt   []byte
```
Comment (abhinavdahiya) [inline pkg/asset/tls/keypair.go]: `private` and `public`;

File: pkg/asset/store.go [hunk 32]
```diff
@@ -16,15 +24,71 @@ type Store interface {
...
+		return errors.Wrapf(err, "failed to unmarshal state file %s", path)
+	}
+	s.stateFileAssets = assets
+	return nil
+}
+
+// GetStateAsset renders the asset object arguments from the state file contents
+// also returns a boolean indicating whether the object was found in the state file or not
+func (s *StoreImpl) GetStateAsset(asset Asset) (bool, error) {
+	logrus.Debugf("Testing state file fetch: %s", reflect.TypeOf(asset).String())
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: maybe edit this log message.

File: pkg/asset/store.go [hunk 33]
```diff
@@ -16,15 +24,71 @@ type Store interface {
 
 // StoreImpl is the implementation of Store.
 type StoreImpl struct {
-	assets map[reflect.Type]Asset
+	assets          map[reflect.Type]Asset
+	stateFileAssets map[string]json.RawMessage
 }
 
+type getTypedObject func(string) (WritableAsset, error)
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: This seems unused.

File: cmd/openshift-install/targets.go [hunk 34]
```diff
@@ -83,6 +87,10 @@ func runTargetCmd(targets ...asset.WritableAsset) func(cmd *cobra.Command, args
 				return errors.Wrapf(err, "failed to write asset (%s) to disk", a.Name())
 			}
 		}
+		err = assetStore.Save(rootOpts.dir)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/targets.go]: can we save state each time we successfully complete a target?

**Comment by abhinavdahiya** [comment]
> The utility function GetStateAsset is left unused in this commit.

Why is this the case? that means we are not using the saved state.

File: pkg/asset/asset_test.go (hunk 20 above)
Comment (rajatchopra) [reply pkg/asset/asset_test.go]: It clashes with the interface function Files. Had to change one of them. Do you prefer to change the function name?

Comment (abhinavdahiya) [reply pkg/asset/asset_test.go]: hmm, lets keep it as is then.

File: cmd/openshift-install/targets.go (hunk 34 above)
Comment (rajatchopra) [reply cmd/openshift-install/targets.go]: Yes. But we use the same StoreImpl, so it will all get overwritten. And worse, doubly computed.

Comment (abhinavdahiya) [reply cmd/openshift-install/targets.go]: sure, lets keep it as is then.

**Comment by rajatchopra** [comment]
> > The utility function GetStateAsset is left unused in this commit.
> 
> Why is this the case? that means we are not using the saved state.

Using the saved state will come in the other PRs when we are ready to do the 3-way merge between state, on-disk and generated asset. I have tested this separately though.
If you think I should just use it now, we can. The following functionality works:
1. Run openshift-install manifests --dir=test
2. Remove the test/{manifests,tectonic} directories
3. Run openshift-install manifests --dir=test, again
   you will not be asked for anything and the manifests/tectonic folders will be restored.
Same with install-config.

**Comment by rajatchopra** [comment]
@abhinavdahiya Put in a feedback-fix commit. Will squash after feedback is resolved.

**Comment by abhinavdahiya** [comment]
> the 3-way merge between state, on-disk and generated asset
That's something we will do in future, but this PR so use the state is saved.

File: pkg/asset/tls/keypair.go [hunk 35]
```diff
@@ -15,9 +15,9 @@ type KeyPairInterface interface {
 
 // KeyPair contains a private key and a public key.
 type KeyPair struct {
-	pvt   []byte
-	pub   []byte
-	files []*asset.File
+	Pvt      []byte
```
Comment (abhinavdahiya) [inline pkg/asset/tls/keypair.go]: do have strong opinion on it not being `Private` and `Public` ?

Comment (rajatchopra) [reply pkg/asset/tls/keypair.go]: I actually wanted to expose these fields, and they clash with the function names. No other reason, I strongly think that Pvt/Pub stinks as compared to Private/Public.

Comment (abhinavdahiya) [reply pkg/asset/tls/keypair.go]: hmm, lets keep it as is then. :/

**Comment by crawford** [comment]
Can you split the export of types and members into a separate commit? It should make review and the history a little easier to understand.

File: pkg/asset/manifests/network-operator.go (hunk 19 above)
Comment (rajatchopra) [reply pkg/asset/manifests/network-operator.go]: Ack. Rolling the struct itself to be private.

**Comment by rajatchopra** [comment]
> Can you split the export of types and members into a separate commit? It should make review and the history a little easier to understand.

Done. Two commits. Both independent of each other. Without the second commit, most of the state file is incomplete (though valid).

**Comment by rajatchopra** [comment]
> > the 3-way merge between state, on-disk and generated asset
> 
> That's something we will do in future, but this PR so use the state is saved.

Added to the first commit - the function is used and now we actually pick up stuff from state file.

File: pkg/asset/store.go [hunk 36]
```diff
@@ -47,10 +108,19 @@ func (s *StoreImpl) fetch(asset Asset, indent string) error {
...
 
-	logrus.Debugf("%sGenerating %s...", indent, asset.Name())
-	err := asset.Generate(parents)
+	// Before generating the asset, look if we have it all ready in the state file
+	// if yes, then use it instead
+	logrus.Debugf("%sFetching asset from state file: %s", indent, reflect.TypeOf(asset).String())
+	ok, err := s.GetStateAsset(asset)
 	if err != nil {
-		return errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+		return errors.Wrapf(err, "failed to unmarshal asset from state file. Remove/fix the state file '%s' and continue..", stateFileName)
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: We should never ask to user to fix the state file. 
this might be enough. `failed to unmashal assets {asset name} from state file{state file name}`

File: pkg/asset/store.go [hunk 37]
```diff
@@ -47,10 +108,19 @@ func (s *StoreImpl) fetch(asset Asset, indent string) error {
...
+	}
+	if !ok {
+		logrus.Debugf("%sAsset not found in state file. Generating %s...", indent, asset.Name())
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: WDYT about this?
```go
if ok {
  // log asset used is from state file
} else {
// log generating assets
// generate asset
}
```

File: pkg/asset/store.go [hunk 38]
```diff
@@ -47,10 +108,19 @@ func (s *StoreImpl) fetch(asset Asset, indent string) error {
 		parents.Add(d)
 	}
 
-	logrus.Debugf("%sGenerating %s...", indent, asset.Name())
-	err := asset.Generate(parents)
+	// Before generating the asset, look if we have it all ready in the state file
+	// if yes, then use it instead
+	logrus.Debugf("%sFetching asset from state file: %s", indent, reflect.TypeOf(asset).String())
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: This should either be dropped or be something like `looking up asset from state file`

File: cmd/openshift-install/targets.go [hunk 39]
```diff
@@ -91,6 +95,10 @@ func runTargetCmd(targets ...asset.WritableAsset) func(cmd *cobra.Command, args
 				return err
 			}
 		}
+		err = assetStore.Save(rootOpts.dir)
+		if err != nil {
+			logrus.Fatalf("Failed to write to state file: %v", err)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/targets.go]: also you can return error here; why Fatal?

File: pkg/asset/store.go (hunk 37 above)
Comment (rajatchopra) [reply pkg/asset/store.go]: That was what I wrote first, then I didn't like the if block with just a log statement :).
Anyhow, redone it, though most of this will get rehashed in a week or so.

**Comment by abhinavdahiya** [comment]
/approve

@staebler @crawford can you take aquick look

**Comment by wking** [comment]
Do we need to poke a hole in this for the not-really-an-asset `cluster` target?  As it stands:

```console
$ openshift-install version
openshift-install v0.2.0-14-g7cc49c8a92a86dd13ac66d1fd0560f679a36d1c0
Terraform v0.11.8
$ openshift-install --log-level=debug cluster
...
DEBUG Looking up asset from state file: *cluster.Cluster 
DEBUG Asset found in state file
```

So I have to manually blow away `.openshift_install_state.json` in order to launch a new cluster.

**Comment by wking** [comment]
> The user directed the asset to be generated, so it should be.

Or maybe they just directed that the asset be written to disk (which it is with the current master).  The thing that's missing is that assets which have important side effects beside being written to disk (e.g. launching the cluster ;) aren't directly addressed by "generate this asset" semantics.

> If the cluster is successfully installed, then the installation process is done. At that point, the state file should be deleted by the installer.

Maybe?  It's not actually clear to me at what point we'd want to remove the state file.  Once `cluster` completes, we'd no longer need the state file for launching clusters, but we might need the Terraform state for removing the bootstrap assets, and we'll need `metadata.json` for `destroy-cluster`.  It feels a bit odd to remove the state file while leaving those.

On the other hand, if you remove the cluster through some other path (e.g. `virsh-cleanup.sh`), then you do want to blow away the cluster state on the next `cluster` call.  So I think the issue is just with the special-side-effects `cluster` "asset", and that we don't need a generic fix.  But I'm not clear on what the `cluster`-specific fix should be ;).

**Comment by rajatchopra** [comment]
@wking @staebler Created PR #476 as a proposed fix. Builds on Matthew's idea that assets that are targeted should always be rendered. Comment over in that PR, if it looks like the right direction.
